dm/pkg/checker/table_structure.go

```diff
@@ -296,88 +361,183 @@ func (c *ShardingTablesChecker) Check(ctx context.Context) *Result {
+		var (
+			instance = firstInstance
+			p        *parser.Parser
+			err      error
+		)
+		defer func() {
+			checkWg.Done()
+			if err != nil {
+				cancel()
+				reMu.Lock()
+				markCheckError(r, err)
+				reMu.Unlock()
+			}
+		}()
+		p, err = dbutil.GetParserForDB(ctx, c.dbs[instance])
 		if err != nil {
-			markCheckError(r, err)
-			r.Extra = fmt.Sprintf("fail to get parser for instance %s on sharding %s", instance, c.name)
-			return r
+			return
 		}
+		for {
+			select {
+			case <-checkCtx.Done():
+				return
+			case checkItem, ok := <-inCh:
+				if !ok {
+					return
+				}
+				table := checkItem.table
+				var err2 error
+				if instance != checkItem.sourceID {
+					instance = checkItem.sourceID
+					p, err2 = dbutil.GetParserForDB(ctx, c.dbs[instance])
+					if err2 != nil {
+						reMu.Lock()
+						r.Extra = fmt.Sprintf("fail to get parser for instance %s on sharding %s", instance, c.targetTableID)
+						reMu.Unlock()
+						err = err2
+						return
+					}
+				}
 
-		for schema, tables := range schemas {
-			for _, table := range tables {
-				statement, err := dbutil.GetCreateTableSQL(ctx, db, schema, table)
-				if err != nil {
+				statement, err2 := dbutil.GetCreateTableSQL(ctx, c.dbs[instance], table.Schema, table.Name)
+				if err2 != nil {
 					// continue if table was deleted when checking
-					if isMySQLError(err, mysql.ErrNoSuchTable) {
+					if isMySQLError(err2, mysql.ErrNoSuchTable) {
 						continue
 					}
-					markCheckError(r, err)
-					r.Extra = fmt.Sprintf("instance %s on sharding %s", instance, c.name)
-					return r
+					err = err2
+					return
 				}
 
-				info, err := dbutil.GetTableInfoBySQL(statement, parser2)
-				if err != nil {
-					markCheckError(r, err)
-					r.Extra = fmt.Sprintf("instance %s on sharding %s", instance, c.name)
-					return r
+				info, err2 := dbutil.GetTableInfoBySQL(statement, p)
+				if err2 != nil {
+					err = errors.Annotatef(err2, "statement: %s", statement)
+					return
 				}
-				stmt, err := parser2.ParseOneStmt(statement, "", "")
-				if err != nil {
-					markCheckError(r, errors.Annotatef(err, "statement %s", statement))
-					r.Extra = fmt.Sprintf("instance %s on sharding %s", instance, c.name)
-					return r
+				stmt, err2 := p.ParseOneStmt(statement, "", "")
+				if err2 != nil {
+					err = errors.Annotatef(err2, "statement: %s", statement)
+					return
 				}
 
 				ctStmt, ok := stmt.(*ast.CreateTableStmt)
 				if !ok {
-					markCheckError(r, errors.Errorf("Expect CreateTableStmt but got %T", stmt))
-					r.Extra = fmt.Sprintf("instance %s on sharding %s", instance, c.name)
-					return r
+					err = errors.Errorf("Expect CreateTableStmt but got %T", stmt)
+					return
 				}
 
 				if c.checkAutoIncrementPrimaryKey {
-					passed := c.checkAutoIncrementKey(instance, schema, table, ctStmt, info, r)
+					passed := c.checkAutoIncrementKey(instance, table.Schema, table.Name, ctStmt, info, r)
 					if !passed {
-						return r
+						return
```

Comment:
if checkAutoIncrementKey not passed, do we need to append a error  in `r.Errors` ? if we return here how checker know this?